{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tutorial 0: Curate Raw CCLE RNA-seq Data\n",
    "\n",
    "In this tutorial we will download publicly available raw RNA-sequencing data, in the form of fastqs, from the Cancer Cell Line Encyclopedia (CCLE) Sequence Read Archive (SRA) repository. \n",
    "\n",
    "This CCLE expression dataset is for use in the demo, and will serve as the basis for generating raw data Quilt packages, and input to Nextflow nf-core pipelines. If available, we encourage you to sub in your in-house datasets in place of the demo CCLE dataset for a more practical use case tailored to your company's platform. \n",
    "\n",
    "Although this demo will be centered around an RNA-seq expression dataset and the corresponding `nf-core/rna-seq` pipeline, the principles and methods underlying the demo are generalizable across -omics datasets. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_columns', None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Curate CCLE metadata\n",
    "\n",
    "First, we want to define a subset of the larger CCLE collection to use for this demo. To define the demo cohort, we used two pieces of publicly available CCLE sample-level metadata: \n",
    "1. [Sequence Read Archive (SRA) Bioproject PRJNA523380](https://www.ncbi.nlm.nih.gov/sra?term=PRJNA523380): sample, sequencing & run metadata\n",
    "2. [cBioportal study](https://www.cbioportal.org/study/summary?id=ccle_broad_2019): detailed biological & clinical sample metadata\n",
    "\n",
    "Both of these raw metadata files are available in the `demo_data/` directory of this repo."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.1 SRA metadata\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>Age</th>\n",
       "      <th>Assay Type</th>\n",
       "      <th>AssemblyName</th>\n",
       "      <th>AvgSpotLen</th>\n",
       "      <th>Bases</th>\n",
       "      <th>BIOMATERIAL_PROVIDER</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>BioSampleModel</th>\n",
       "      <th>Bytes</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>Center Name</th>\n",
       "      <th>Consent</th>\n",
       "      <th>DATASTORE filetype</th>\n",
       "      <th>DATASTORE provider</th>\n",
       "      <th>DATASTORE region</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease_stage</th>\n",
       "      <th>ETHNICITY</th>\n",
       "      <th>Experiment</th>\n",
       "      <th>Instrument</th>\n",
       "      <th>isolate</th>\n",
       "      <th>Library Name</th>\n",
       "      <th>LibraryLayout</th>\n",
       "      <th>LibrarySelection</th>\n",
       "      <th>LibrarySource</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Platform</th>\n",
       "      <th>ReleaseDate</th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>sex</th>\n",
       "      <th>SRA Study</th>\n",
       "      <th>tissue</th>\n",
       "      <th>create_date</th>\n",
       "      <th>version</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR8670667</td>\n",
       "      <td>81.0</td>\n",
       "      <td>WGS</td>\n",
       "      <td>GCA_000001405.13</td>\n",
       "      <td>202</td>\n",
       "      <td>128791662980</td>\n",
       "      <td>ATCC:T24</td>\n",
       "      <td>PRJNA523380</td>\n",
       "      <td>SAMN10987918</td>\n",
       "      <td>Human</td>\n",
       "      <td>68905947744</td>\n",
       "      <td>T24</td>\n",
       "      <td>BROAD INSTITUTE</td>\n",
       "      <td>public</td>\n",
       "      <td>bam,run.zq,sra</td>\n",
       "      <td>gs,ncbi,s3</td>\n",
       "      <td>gs.US,ncbi.public,s3.us-east-1</td>\n",
       "      <td>carcinoma (transitional_cell_carcinoma)</td>\n",
       "      <td>primary</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>SRX5466713</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>cell line</td>\n",
       "      <td>WGS-T24_URINARY_TRACT</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>size fractionation</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2019-03-27T00:00:00Z</td>\n",
       "      <td>T24_URINARY_TRACT</td>\n",
       "      <td>cell culture</td>\n",
       "      <td>female</td>\n",
       "      <td>SRP186687</td>\n",
       "      <td>urinary_tract</td>\n",
       "      <td>2019-03-05T22:40:00Z</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR8670668</td>\n",
       "      <td>81.0</td>\n",
       "      <td>WGS</td>\n",
       "      <td>GCA_000001405.13</td>\n",
       "      <td>202</td>\n",
       "      <td>102955983372</td>\n",
       "      <td>NIBRI</td>\n",
       "      <td>PRJNA523380</td>\n",
       "      <td>SAMN10988334</td>\n",
       "      <td>Human</td>\n",
       "      <td>46845960495</td>\n",
       "      <td>SW948</td>\n",
       "      <td>BROAD INSTITUTE</td>\n",
       "      <td>public</td>\n",
       "      <td>bam,run.zq,sra</td>\n",
       "      <td>gs,ncbi,s3</td>\n",
       "      <td>gs.US,ncbi.public,s3.us-east-1</td>\n",
       "      <td>carcinoma (adenocarcinoma)</td>\n",
       "      <td>primary</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>SRX5466712</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>cell line</td>\n",
       "      <td>WGS-SW948_LARGE_INTESTINE</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>size fractionation</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2019-03-27T00:00:00Z</td>\n",
       "      <td>SW948_LARGE_INTESTINE</td>\n",
       "      <td>cell culture</td>\n",
       "      <td>female</td>\n",
       "      <td>SRP186687</td>\n",
       "      <td>large_intestine</td>\n",
       "      <td>2019-03-05T17:38:00Z</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR8670669</td>\n",
       "      <td>53.0</td>\n",
       "      <td>WGS</td>\n",
       "      <td>GCA_000001405.13</td>\n",
       "      <td>202</td>\n",
       "      <td>117381684900</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PRJNA523380</td>\n",
       "      <td>SAMN10987793</td>\n",
       "      <td>Human</td>\n",
       "      <td>60335385048</td>\n",
       "      <td>SW 900</td>\n",
       "      <td>BROAD INSTITUTE</td>\n",
       "      <td>public</td>\n",
       "      <td>bam,run.zq,sra</td>\n",
       "      <td>gs,ncbi,s3</td>\n",
       "      <td>gs.US,ncbi.public,s3.us-east-1</td>\n",
       "      <td>carcinoma (squamous_cell_carcinoma)</td>\n",
       "      <td>primary</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>SRX5466711</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>cell line</td>\n",
       "      <td>WGS-SW900_LUNG</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>size fractionation</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2019-03-27T00:00:00Z</td>\n",
       "      <td>SW900_LUNG</td>\n",
       "      <td>cell culture</td>\n",
       "      <td>male</td>\n",
       "      <td>SRP186687</td>\n",
       "      <td>lung</td>\n",
       "      <td>2019-03-05T19:22:00Z</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR8670670</td>\n",
       "      <td>59.0</td>\n",
       "      <td>WGS</td>\n",
       "      <td>GCA_000001405.13</td>\n",
       "      <td>202</td>\n",
       "      <td>136392701588</td>\n",
       "      <td>GNF</td>\n",
       "      <td>PRJNA523380</td>\n",
       "      <td>SAMN10988339</td>\n",
       "      <td>Human</td>\n",
       "      <td>76410262878</td>\n",
       "      <td>SW579</td>\n",
       "      <td>BROAD INSTITUTE</td>\n",
       "      <td>public</td>\n",
       "      <td>bam,run.zq,sra</td>\n",
       "      <td>gs,ncbi,s3</td>\n",
       "      <td>gs.US,ncbi.public,s3.us-east-1</td>\n",
       "      <td>carcinoma (anaplastic_carcinoma)</td>\n",
       "      <td>primary</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>SRX5466710</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>cell line</td>\n",
       "      <td>WGS-SW579_THYROID</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>size fractionation</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2019-03-27T00:00:00Z</td>\n",
       "      <td>SW579_THYROID</td>\n",
       "      <td>cell culture</td>\n",
       "      <td>male</td>\n",
       "      <td>SRP186687</td>\n",
       "      <td>thyroid</td>\n",
       "      <td>2019-03-05T21:09:00Z</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR8670671</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WGS</td>\n",
       "      <td>GCA_000001405.13</td>\n",
       "      <td>202</td>\n",
       "      <td>111933931548</td>\n",
       "      <td>JCRB:TE-1</td>\n",
       "      <td>PRJNA523380</td>\n",
       "      <td>SAMN10988294</td>\n",
       "      <td>Human</td>\n",
       "      <td>55384403411</td>\n",
       "      <td>TE-1</td>\n",
       "      <td>BROAD INSTITUTE</td>\n",
       "      <td>public</td>\n",
       "      <td>bam,run.zq,sra</td>\n",
       "      <td>gs,ncbi,s3</td>\n",
       "      <td>gs.US,ncbi.public,s3.us-east-1</td>\n",
       "      <td>carcinoma (squamous_cell_carcinoma)</td>\n",
       "      <td>primary</td>\n",
       "      <td>Asian</td>\n",
       "      <td>SRX5466709</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>cell line</td>\n",
       "      <td>WGS-TE1_OESOPHAGUS</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>size fractionation</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2019-03-27T00:00:00Z</td>\n",
       "      <td>TE1_OESOPHAGUS</td>\n",
       "      <td>cell culture</td>\n",
       "      <td>male</td>\n",
       "      <td>SRP186687</td>\n",
       "      <td>oesophagus</td>\n",
       "      <td>2019-03-05T18:52:00Z</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Run   Age Assay Type      AssemblyName  AvgSpotLen         Bases  \\\n",
       "0  SRR8670667  81.0        WGS  GCA_000001405.13         202  128791662980   \n",
       "1  SRR8670668  81.0        WGS  GCA_000001405.13         202  102955983372   \n",
       "2  SRR8670669  53.0        WGS  GCA_000001405.13         202  117381684900   \n",
       "3  SRR8670670  59.0        WGS  GCA_000001405.13         202  136392701588   \n",
       "4  SRR8670671   NaN        WGS  GCA_000001405.13         202  111933931548   \n",
       "\n",
       "  BIOMATERIAL_PROVIDER   BioProject     BioSample BioSampleModel        Bytes  \\\n",
       "0             ATCC:T24  PRJNA523380  SAMN10987918          Human  68905947744   \n",
       "1                NIBRI  PRJNA523380  SAMN10988334          Human  46845960495   \n",
       "2                  NaN  PRJNA523380  SAMN10987793          Human  60335385048   \n",
       "3                  GNF  PRJNA523380  SAMN10988339          Human  76410262878   \n",
       "4            JCRB:TE-1  PRJNA523380  SAMN10988294          Human  55384403411   \n",
       "\n",
       "  cell_line      Center Name Consent DATASTORE filetype DATASTORE provider  \\\n",
       "0       T24  BROAD INSTITUTE  public     bam,run.zq,sra         gs,ncbi,s3   \n",
       "1     SW948  BROAD INSTITUTE  public     bam,run.zq,sra         gs,ncbi,s3   \n",
       "2    SW 900  BROAD INSTITUTE  public     bam,run.zq,sra         gs,ncbi,s3   \n",
       "3     SW579  BROAD INSTITUTE  public     bam,run.zq,sra         gs,ncbi,s3   \n",
       "4      TE-1  BROAD INSTITUTE  public     bam,run.zq,sra         gs,ncbi,s3   \n",
       "\n",
       "                 DATASTORE region                                  disease  \\\n",
       "0  gs.US,ncbi.public,s3.us-east-1  carcinoma (transitional_cell_carcinoma)   \n",
       "1  gs.US,ncbi.public,s3.us-east-1               carcinoma (adenocarcinoma)   \n",
       "2  gs.US,ncbi.public,s3.us-east-1      carcinoma (squamous_cell_carcinoma)   \n",
       "3  gs.US,ncbi.public,s3.us-east-1         carcinoma (anaplastic_carcinoma)   \n",
       "4  gs.US,ncbi.public,s3.us-east-1      carcinoma (squamous_cell_carcinoma)   \n",
       "\n",
       "  disease_stage  ETHNICITY  Experiment           Instrument    isolate  \\\n",
       "0       primary  Caucasian  SRX5466713  Illumina HiSeq 2000  cell line   \n",
       "1       primary  Caucasian  SRX5466712  Illumina HiSeq 2000  cell line   \n",
       "2       primary  Caucasian  SRX5466711  Illumina HiSeq 2000  cell line   \n",
       "3       primary  Caucasian  SRX5466710  Illumina HiSeq 2000  cell line   \n",
       "4       primary      Asian  SRX5466709  Illumina HiSeq 2000  cell line   \n",
       "\n",
       "                Library Name LibraryLayout    LibrarySelection LibrarySource  \\\n",
       "0      WGS-T24_URINARY_TRACT        PAIRED  size fractionation       GENOMIC   \n",
       "1  WGS-SW948_LARGE_INTESTINE        PAIRED  size fractionation       GENOMIC   \n",
       "2             WGS-SW900_LUNG        PAIRED  size fractionation       GENOMIC   \n",
       "3          WGS-SW579_THYROID        PAIRED  size fractionation       GENOMIC   \n",
       "4         WGS-TE1_OESOPHAGUS        PAIRED  size fractionation       GENOMIC   \n",
       "\n",
       "       Organism  Platform           ReleaseDate            Sample Name  \\\n",
       "0  Homo sapiens  ILLUMINA  2019-03-27T00:00:00Z      T24_URINARY_TRACT   \n",
       "1  Homo sapiens  ILLUMINA  2019-03-27T00:00:00Z  SW948_LARGE_INTESTINE   \n",
       "2  Homo sapiens  ILLUMINA  2019-03-27T00:00:00Z             SW900_LUNG   \n",
       "3  Homo sapiens  ILLUMINA  2019-03-27T00:00:00Z          SW579_THYROID   \n",
       "4  Homo sapiens  ILLUMINA  2019-03-27T00:00:00Z         TE1_OESOPHAGUS   \n",
       "\n",
       "    sample_type     sex  SRA Study           tissue           create_date  \\\n",
       "0  cell culture  female  SRP186687    urinary_tract  2019-03-05T22:40:00Z   \n",
       "1  cell culture  female  SRP186687  large_intestine  2019-03-05T17:38:00Z   \n",
       "2  cell culture    male  SRP186687             lung  2019-03-05T19:22:00Z   \n",
       "3  cell culture    male  SRP186687          thyroid  2019-03-05T21:09:00Z   \n",
       "4  cell culture    male  SRP186687       oesophagus  2019-03-05T18:52:00Z   \n",
       "\n",
       "   version  \n",
       "0      1.0  \n",
       "1      1.0  \n",
       "2      1.0  \n",
       "3      1.0  \n",
       "4      1.0  "
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load SRA CCLE metadata\n",
    "# rows are CCLE samples, columns are metadata fields\n",
    "sra = pd.read_csv(\"./demo_data/sample_metadata/downloads/CCLE_SraRunTable.txt\")\n",
    "sra.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "BioProject\n",
       "PRJNA523380    4550\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# confirm all CCLE samples are part of the same BioProject -- yes!\n",
    "sra[\"BioProject\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Run\n",
       "SRR8670667    1\n",
       "SRR8633850    1\n",
       "SRR8633856    1\n",
       "SRR8633855    1\n",
       "SRR8633854    1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# each sample+assay is its own run\n",
    "sra[\"Run\"].value_counts().head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Assay Type\n",
       "RNA-Seq             1019\n",
       "Targeted-Capture     976\n",
       "AMPLICON             972\n",
       "Bisulfite-Seq        928\n",
       "WGS                  329\n",
       "WXS                  326\n",
       "Name: count, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(1019, 37)"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# subset to samples profiled with RNA-seq only (n=1,019)\n",
    "display(sra[\"Assay Type\"].value_counts())\n",
    "sra_rna = sra.loc[sra[\"Assay Type\"] == \"RNA-Seq\", :]\n",
    "sra_rna.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.2 Merge cBioportal & SRA metadata together \n",
    "\n",
    "SRA metadata has very detailed sequencing & sample processing information. To compliment this, we will add detailed clinical & biological information from the corresponding CCLE cBioportal study."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Study ID</th>\n",
       "      <th>Patient ID</th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>Age</th>\n",
       "      <th>Annotation Source</th>\n",
       "      <th>Cancer Type</th>\n",
       "      <th>Cancer Type Detailed</th>\n",
       "      <th>Cell Line Source</th>\n",
       "      <th>Characteristics</th>\n",
       "      <th>DepMap ID</th>\n",
       "      <th>Disease Ontology</th>\n",
       "      <th>Doubling Time (hrs)</th>\n",
       "      <th>Doubling Time From Vendor</th>\n",
       "      <th>Ethnicity Category</th>\n",
       "      <th>Fraction Genome Altered</th>\n",
       "      <th>Freezing Medium</th>\n",
       "      <th>Genome Doublings</th>\n",
       "      <th>Geographic Distribution</th>\n",
       "      <th>Growth Medium</th>\n",
       "      <th>Histology</th>\n",
       "      <th>Hist Subtype 1</th>\n",
       "      <th>Hist Subtype 2</th>\n",
       "      <th>Hist Subtype 3</th>\n",
       "      <th>Life Stage</th>\n",
       "      <th>Lineage</th>\n",
       "      <th>Lineage Molecular Subtype</th>\n",
       "      <th>Lineage Subtype</th>\n",
       "      <th>Lineage Sub-subtype</th>\n",
       "      <th>Mutation Count</th>\n",
       "      <th>Mutation Rate</th>\n",
       "      <th>Name</th>\n",
       "      <th>Oncotree Code</th>\n",
       "      <th>Pathologist Annotation</th>\n",
       "      <th>Ploidy</th>\n",
       "      <th>Primary Tumor Site</th>\n",
       "      <th>Proteomics 10-Plex ID</th>\n",
       "      <th>Proteomics TMT label</th>\n",
       "      <th>Purity</th>\n",
       "      <th>Race Category</th>\n",
       "      <th>Number of Samples Per Patient</th>\n",
       "      <th>Sample Type</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Site of Finding</th>\n",
       "      <th>Site Subtype 1</th>\n",
       "      <th>Site Subtype 2</th>\n",
       "      <th>Site Subtype 3</th>\n",
       "      <th>Subtype</th>\n",
       "      <th>Supplements</th>\n",
       "      <th>TMB (nonsynonymous)</th>\n",
       "      <th>Tumor Type</th>\n",
       "      <th>Type Refined</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ccle_broad_2019</td>\n",
       "      <td>127399_SOFT_TISSUE</td>\n",
       "      <td>127399_SOFT_TISSUE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Soft Tissue Sarcoma</td>\n",
       "      <td>Synovial Sarcoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACH-001270</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>soft_tissue</td>\n",
       "      <td>NaN</td>\n",
       "      <td>synovial_sarcoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>163.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SYNS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Synovial</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.533333</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ccle_broad_2019</td>\n",
       "      <td>1321N1_CENTRAL_NERVOUS_SYSTEM</td>\n",
       "      <td>1321N1_CENTRAL_NERVOUS_SYSTEM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CCLE</td>\n",
       "      <td>Glioma</td>\n",
       "      <td>Astrocytoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACH-001000</td>\n",
       "      <td>Brain Cancer</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>0.5648</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Glioma</td>\n",
       "      <td>astrocytoma</td>\n",
       "      <td>NS</td>\n",
       "      <td>NS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>central_nervous_system</td>\n",
       "      <td>NaN</td>\n",
       "      <td>glioma</td>\n",
       "      <td>astrocytoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1321N1</td>\n",
       "      <td>ASTR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.45</td>\n",
       "      <td>Central_Nervous_System</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.98</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>Primary</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>brain</td>\n",
       "      <td>NS</td>\n",
       "      <td>NS</td>\n",
       "      <td>Astrocytoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>glioma</td>\n",
       "      <td>glioma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ccle_broad_2019</td>\n",
       "      <td>143B_BONE</td>\n",
       "      <td>143B_BONE</td>\n",
       "      <td>13.0</td>\n",
       "      <td>CCLE</td>\n",
       "      <td>Bone Cancer</td>\n",
       "      <td>Osteosarcoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACH-001001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>0.5014</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Osteosarcoma</td>\n",
       "      <td>NS</td>\n",
       "      <td>NS</td>\n",
       "      <td>NS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>bone</td>\n",
       "      <td>NaN</td>\n",
       "      <td>osteosarcoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>143B</td>\n",
       "      <td>OS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.61</td>\n",
       "      <td>Bone</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.98</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>1</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NS</td>\n",
       "      <td>NS</td>\n",
       "      <td>NS</td>\n",
       "      <td>Osteosarcoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>osteosarcoma</td>\n",
       "      <td>osteosarcoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ccle_broad_2019</td>\n",
       "      <td>201T_LUNG</td>\n",
       "      <td>201T_LUNG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Non-Small Cell Lung Cancer</td>\n",
       "      <td>Lung Adenocarcinoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACH-002089</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lung</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NSCLC</td>\n",
       "      <td>NSCLC_adenocarcinoma</td>\n",
       "      <td>290.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Non-Small Cell Lung Cancer (NSCLC), Adenocarci...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.100000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lung_NSC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ccle_broad_2019</td>\n",
       "      <td>22RV1_PROSTATE</td>\n",
       "      <td>22RV1_PROSTATE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CCLE</td>\n",
       "      <td>Prostate Cancer</td>\n",
       "      <td>Prostate Adenocarcinoma</td>\n",
       "      <td>ATCC</td>\n",
       "      <td>Adherent epithelial</td>\n",
       "      <td>ACH-000956</td>\n",
       "      <td>prostate_cancer</td>\n",
       "      <td>58.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>0.2002</td>\n",
       "      <td>5%DMSO</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RPMI-1640+10% FBS</td>\n",
       "      <td>Carcinoma</td>\n",
       "      <td>NS</td>\n",
       "      <td>NS</td>\n",
       "      <td>NS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>prostate</td>\n",
       "      <td>NaN</td>\n",
       "      <td>prostate_adenocarcinoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2186.0</td>\n",
       "      <td>539.830868</td>\n",
       "      <td>22Rv1</td>\n",
       "      <td>PRAD</td>\n",
       "      <td>Prostate:Carcinoma</td>\n",
       "      <td>2.13</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>24.0</td>\n",
       "      <td>126</td>\n",
       "      <td>0.99</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NS</td>\n",
       "      <td>NS</td>\n",
       "      <td>NS</td>\n",
       "      <td>Adenocarcinoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>73.700000</td>\n",
       "      <td>prostate</td>\n",
       "      <td>prostate</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Study ID                     Patient ID  \\\n",
       "0  ccle_broad_2019             127399_SOFT_TISSUE   \n",
       "1  ccle_broad_2019  1321N1_CENTRAL_NERVOUS_SYSTEM   \n",
       "2  ccle_broad_2019                      143B_BONE   \n",
       "3  ccle_broad_2019                      201T_LUNG   \n",
       "4  ccle_broad_2019                 22RV1_PROSTATE   \n",
       "\n",
       "                       Sample ID   Age Annotation Source  \\\n",
       "0             127399_SOFT_TISSUE   NaN               NaN   \n",
       "1  1321N1_CENTRAL_NERVOUS_SYSTEM   NaN              CCLE   \n",
       "2                      143B_BONE  13.0              CCLE   \n",
       "3                      201T_LUNG   NaN               NaN   \n",
       "4                 22RV1_PROSTATE   NaN              CCLE   \n",
       "\n",
       "                  Cancer Type     Cancer Type Detailed Cell Line Source  \\\n",
       "0         Soft Tissue Sarcoma         Synovial Sarcoma              NaN   \n",
       "1                      Glioma              Astrocytoma              NaN   \n",
       "2                 Bone Cancer             Osteosarcoma              NaN   \n",
       "3  Non-Small Cell Lung Cancer      Lung Adenocarcinoma              NaN   \n",
       "4             Prostate Cancer  Prostate Adenocarcinoma             ATCC   \n",
       "\n",
       "       Characteristics   DepMap ID Disease Ontology  Doubling Time (hrs)  \\\n",
       "0                  NaN  ACH-001270              NaN                  NaN   \n",
       "1                  NaN  ACH-001000     Brain Cancer                  NaN   \n",
       "2                  NaN  ACH-001001              NaN                  NaN   \n",
       "3                  NaN  ACH-002089              NaN                  NaN   \n",
       "4  Adherent epithelial  ACH-000956  prostate_cancer                 58.4   \n",
       "\n",
       "  Doubling Time From Vendor Ethnicity Category  Fraction Genome Altered  \\\n",
       "0                       NaN                NaN                      NaN   \n",
       "1                       NaN          Caucasian                   0.5648   \n",
       "2                       NaN          Caucasian                   0.5014   \n",
       "3                       NaN                NaN                      NaN   \n",
       "4                       NaN          Caucasian                   0.2002   \n",
       "\n",
       "  Freezing Medium  Genome Doublings Geographic Distribution  \\\n",
       "0             NaN               NaN                     NaN   \n",
       "1             NaN               2.0                     NaN   \n",
       "2             NaN               2.0                     NaN   \n",
       "3             NaN               NaN                     NaN   \n",
       "4          5%DMSO               0.0                     NaN   \n",
       "\n",
       "       Growth Medium     Histology Hist Subtype 1 Hist Subtype 2  \\\n",
       "0                NaN           NaN            NaN            NaN   \n",
       "1                NaN        Glioma    astrocytoma             NS   \n",
       "2                NaN  Osteosarcoma             NS             NS   \n",
       "3                NaN           NaN            NaN            NaN   \n",
       "4  RPMI-1640+10% FBS     Carcinoma             NS             NS   \n",
       "\n",
       "  Hist Subtype 3 Life Stage                 Lineage Lineage Molecular Subtype  \\\n",
       "0            NaN        NaN             soft_tissue                       NaN   \n",
       "1             NS        NaN  central_nervous_system                       NaN   \n",
       "2             NS        NaN                    bone                       NaN   \n",
       "3            NaN        NaN                    lung                       NaN   \n",
       "4             NS        NaN                prostate                       NaN   \n",
       "\n",
       "           Lineage Subtype   Lineage Sub-subtype  Mutation Count  \\\n",
       "0         synovial_sarcoma                   NaN           163.0   \n",
       "1                   glioma           astrocytoma             NaN   \n",
       "2             osteosarcoma                   NaN             NaN   \n",
       "3                    NSCLC  NSCLC_adenocarcinoma           290.0   \n",
       "4  prostate_adenocarcinoma                   NaN          2186.0   \n",
       "\n",
       "   Mutation Rate    Name Oncotree Code Pathologist Annotation  Ploidy  \\\n",
       "0            NaN     NaN          SYNS                    NaN     NaN   \n",
       "1            NaN  1321N1          ASTR                    NaN    3.45   \n",
       "2            NaN    143B            OS                    NaN    3.61   \n",
       "3            NaN     NaN          LUAD                    NaN     NaN   \n",
       "4     539.830868   22Rv1          PRAD     Prostate:Carcinoma    2.13   \n",
       "\n",
       "       Primary Tumor Site  Proteomics 10-Plex ID Proteomics TMT label  Purity  \\\n",
       "0                     NaN                    NaN                  NaN     NaN   \n",
       "1  Central_Nervous_System                    NaN                  NaN    0.98   \n",
       "2                    Bone                    NaN                  NaN    0.98   \n",
       "3                     NaN                    NaN                  NaN     NaN   \n",
       "4                Prostate                   24.0                  126    0.99   \n",
       "\n",
       "  Race Category  Number of Samples Per Patient Sample Type     Sex  \\\n",
       "0           NaN                              1         NaN     NaN   \n",
       "1           NaN                              1     Primary     NaN   \n",
       "2     Caucasian                              1     Primary  Female   \n",
       "3           NaN                              1         NaN     NaN   \n",
       "4           NaN                              1     Primary    Male   \n",
       "\n",
       "  Site of Finding Site Subtype 1 Site Subtype 2 Site Subtype 3  \\\n",
       "0             NaN            NaN            NaN            NaN   \n",
       "1             NaN          brain             NS             NS   \n",
       "2             NaN             NS             NS             NS   \n",
       "3             NaN            NaN            NaN            NaN   \n",
       "4             NaN             NS             NS             NS   \n",
       "\n",
       "                                             Subtype Supplements  \\\n",
       "0                                           Synovial         NaN   \n",
       "1                                        Astrocytoma         NaN   \n",
       "2                                       Osteosarcoma         NaN   \n",
       "3  Non-Small Cell Lung Cancer (NSCLC), Adenocarci...         NaN   \n",
       "4                                     Adenocarcinoma         NaN   \n",
       "\n",
       "   TMB (nonsynonymous)    Tumor Type  Type Refined  \n",
       "0             5.533333           NaN           NaN  \n",
       "1                  NaN        glioma        glioma  \n",
       "2                  NaN  osteosarcoma  osteosarcoma  \n",
       "3            10.100000           NaN      lung_NSC  \n",
       "4            73.700000      prostate      prostate  "
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cbio = pd.read_csv(\"./demo_data/sample_metadata/downloads/cbio_ccle_broad_2019_clinical_data.tsv\",\n",
    "                   sep = \"\\t\"\n",
    "                  )\n",
    "cbio.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Sample Name\n",
       "True    1019\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# confirm SRA Sample names are present in cBio metadata -- yes!\n",
    "sra_rna[\"Sample Name\"].isin(cbio[\"Sample ID\"]).value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge cBio & sra metadata \n",
    "cbio.set_index(\"Sample ID\", inplace=True)\n",
    "cbio = cbio.drop(columns=\"Age\")\n",
    "meta = sra_rna.merge(cbio, left_on = \"Sample Name\", right_index = True, how = \"left\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Metadata cleanup\n",
    "\n",
    "Now that we have a big table of comprehensive metadata, we want to harmonize inconsistencies & column header format, as well as drop repetitive columns with redundant information on the sample (mainly an artifact of pulling metadata fro two sources, there will be some overlap in information).\n",
    "\n",
    "Having clean metadata is essential to making your data findable! Harmonizing spelling, formatting etc. upfront in the early stages of a project can be extremely powerful & time-saving to having clean covariates when it comes time for analysis. But more on that in Tutorial #1 where we cover metadata schemas & workflows for Quilt packages!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.1 Harmonize freezing medium\n",
    "\n",
    "Ugly mix of character types & naming inconsistencies decribing the same sample feature. \n",
    "\n",
    "For example, `5%DMSO` vs. `5% DMSO` vs. `5%DMSo` vs. `5` vs. `5 % DMSO`. \n",
    "\n",
    "The metadata lists the same 5% DMSO variable 5 different ways! Lets harmonize this..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Freezing Medium\n",
       "5%DMSO        242\n",
       "5% DMSO       221\n",
       "10%DMSO        45\n",
       "10% DMSO        9\n",
       "?               8\n",
       "5%DMSo          2\n",
       "5               1\n",
       "5-7.5%DMSO      1\n",
       "5 % DMSO        1\n",
       "0.05            1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Oof, thats ugly\n",
    "meta[\"Freezing Medium\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta[\"Freezing Medium\"] = meta[\"Freezing Medium\"].str.replace(\"5%DMSO\", \"5% DMSO\")\n",
    "meta[\"Freezing Medium\"] = meta[\"Freezing Medium\"].str.replace(\"5%DMSo\", \"5% DMSO\")\n",
    "meta[\"Freezing Medium\"] = meta[\"Freezing Medium\"].str.replace(\"5 % DMSO\", \"5% DMSO\")\n",
    "meta[\"Freezing Medium\"] = meta[\"Freezing Medium\"].str.replace(\"10%DMSO\", \"10% DMSO\")\n",
    "meta[\"Freezing Medium\"] = meta[\"Freezing Medium\"].str.replace(\"5\", \"5% DMSO\")\n",
    "meta[\"Freezing Medium\"] = meta[\"Freezing Medium\"].str.replace(\"5% DMSO% DMSO\", \"5% DMSO\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "# replace question mark with np.nan \n",
    "# to be consistent with unknown var used in other columns\n",
    "meta[\"Freezing Medium\"] = meta[\"Freezing Medium\"].replace(\"?\", np.nan)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Freezing Medium\n",
       "5% DMSO              467\n",
       "10% DMSO              54\n",
       "5% DMSO-7.5% DMSO      1\n",
       "0.05% DMSO             1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ah, looks much better :) \n",
    "meta[\"Freezing Medium\"].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.2 Drop redundant columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# define columns to remove\n",
    "# these columns contain the same information as another column\n",
    "remove = [\"Doubling Time From Vendor\", \n",
    "          \"Ethnicity Category\",\n",
    "          \"Sex\",\n",
    "          \"Hist Subtype 3\",\n",
    "          \"Race Category\",\n",
    "          \"Number of Samples Per Patient\",\n",
    "          \"Sample Type\",\n",
    "          \"Site Subtype 3\",\n",
    "          \"Tumor Type\",\n",
    "          \"Type Refined\"\n",
    "         ]\n",
    "len(remove)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1019, 76)"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta = meta.loc[:, ~meta.columns.isin(remove)]\n",
    "meta.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.3 Harmonize column formatting\n",
    "\n",
    "Use CamelCase for column headers - make column headers with same format between SRA & cBioportal. This will make our lives easier when defining the metadata schema & workflow, and also...looks nicer!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "col_update = {\"Assay Type\": \"AssayType\",\n",
    "              \"BIOMATERIAL_PROVIDER\": \"BiomaterialProvider\",\n",
    "              \"cell_line\": \"CellLine\",\n",
    "              \"Center Name\": \"CenterName\",\n",
    "              \"DATASTORE filetype\": \"DataStoreFileType\",\n",
    "              \"DATASTORE provider\": \"DataStoreProvider\",\n",
    "              \"DATASTORE region\": \"DataStoreRegion\",\n",
    "              \"disease\": \"Disease\",\n",
    "              \"disease_stage\": \"DiseaseStage\",\n",
    "              \"ETHNICITY\": \"Ethnicity\",\n",
    "              \"isolate\": \"Isolate\",\n",
    "              \"Library Name\": \"LibraryName\",\n",
    "              \"Sample Name\": \"SampleName\",\n",
    "              \"sample_type\": \"SampleType\",\n",
    "              \"sex\": \"Sex\",\n",
    "              \"SRA Study\": \"SRAStudy\",\n",
    "              \"tissue\": \"Tissue\",\n",
    "              \"create_date\": \"CreateDate\",\n",
    "              \"create_date_batch\": \"CreateDateBatch\",\n",
    "              \"version\": \"Version\",\n",
    "              \"Study ID\": \"StudyID\",\n",
    "              \"Patient ID\": \"PatientID\",\n",
    "              \"Annotation Source\": \"AnnotationSource\",\n",
    "              \"Cancer Type\": \"CancerType\",\n",
    "              \"Cancer Type Detailed\": \"CancerTypeDetailed\",\n",
    "              \"Cell Line Source\": \"CellLineSource\",\n",
    "              \"DepMap ID\": \"DepMapID\",\n",
    "              \"Disease Ontology\": \"DiseaseOntology\",\n",
    "              \"Doubling Time (hrs)\": \"DoublingTimeHrs\",\n",
    "              \"Fraction Genome Altered\": \"FractionGenomeAltered\",\n",
    "              \"Freezing Medium\": \"FreezingMedium\",\n",
    "              \"Genome Doublings\": \"GenomeDoublings\",\n",
    "              \"Geographic Distribution\": \"GeographicDistribution\",\n",
    "              \"Growth Medium\": \"GrowthMedium\",\n",
    "              \"Hist Subtype 1\": \"HistologySubtype1\",\n",
    "              \"Hist Subtype 2\": \"HistologySubtype2\",\n",
    "              \"Life Stage\": \"LifeStage\",\n",
    "              \"Lineage Molecular Subtype\": \"LineageMolecularSubtype\",\n",
    "              \"Lineage Subtype\": \"LineageSubtype\",\n",
    "              \"Lineage Sub-subtype\": \"LineageSubSubtype\",\n",
    "              \"Mutation Count\": \"MutationCount\",\n",
    "              \"Mutation Rate\": \"MutationRate\",\n",
    "              \"Name\": \"CellLineNickName\",\n",
    "              \"Oncotree Code\": \"OncotreeCode\",\n",
    "              \"Pathologist Annotation\": \"PathologistAnnotation\",\n",
    "              \"Primary Tumor Site\": \"PrimaryTumorSite\",\n",
    "              \"Proteomics 10-Plex ID\": \"Proteomics10PlexID\",\n",
    "              \"Proteomics TMT label\": \"ProteomicsTMTLabel\",\n",
    "              \"Site of Finding\": \"SiteOfFinding\",\n",
    "              \"Site Subtype 1\": \"SiteSubtype1\",\n",
    "              \"Site Subtype 2\": \"SiteSubtype2\",\n",
    "              \"TMB (nonsynonymous)\": \"TMBNonSynonymous\"              \n",
    "             }\n",
    "meta = meta.rename(columns=col_update)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.4 Replace NA's with Unknown\n",
    "\n",
    "The metadata schema in Tutorial 1 has a set of required fields that must be filled for each sample. NA's are not considered \"filled\" values. To get around this, we can replace NA's with the string \"Unknown\", which counts as a valid value in our schema. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "# list of required fields, as stated in metadata schema in Tutorial 1\n",
    "required = [\"Run\",\n",
    "            \"SRAStudy\",\n",
    "            \"AssayType\",\n",
    "            \"AvgSpotLen\",\n",
    "            \"BiomaterialProvider\",\n",
    "            \"BioProject\",\n",
    "            \"BioSample\",\n",
    "            \"BioSampleModel\",\n",
    "            \"CellLine\",\n",
    "            \"CenterName\",\n",
    "            \"Consent\",\n",
    "            \"Disease\",\n",
    "            \"DiseaseStage\",\n",
    "            \"Ethnicity\",\n",
    "            \"Experiment\",\n",
    "            \"Instrument\",\n",
    "            \"Isolate\",\n",
    "            \"LibraryName\",\n",
    "            \"LibraryLayout\",\n",
    "            \"LibrarySelection\",\n",
    "            \"LibrarySource\",\n",
    "            \"Organism\",\n",
    "            \"Platform\",\n",
    "            \"ReleaseDate\",\n",
    "            \"SampleName\",\n",
    "            \"SampleType\",\n",
    "            \"Sex\",\n",
    "            \"Tissue\",\n",
    "            \"CreateDate\",\n",
    "            \"Version\"\n",
    "    ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta[required] = meta[required].fillna(value=\"Unknown\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.5 Create explicit SampleId field\n",
    "\n",
    "`Run` is used as the sample ID for studies on SRA. However, this nomenclature is slightly confusing with a sequencing or pipeline run. To circumvent this, we will add an explicit `SampleID` field in the metadata table. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta.insert(0, \"SampleID\", meta[\"Run\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Define CCLE sample subset for demo\n",
    "\n",
    "Now that we have the sample metadata all cleaned up, we need to define a subset of samples to use in our demo. To best mirror \"real world\" workflows in the lab, where samples are processed & sequenced on a rolling basis - we decided to select samples that span multiple sequencing/processing batches.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3.1 Create sample batch labels\n",
    "\n",
    "Use date created for batches."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CreateDate</th>\n",
       "      <th>nSamples</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2019-02-23T22:52:00Z</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2019-02-23T23:37:00Z</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2019-02-23T22:58:00Z</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2019-02-23T23:01:00Z</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2019-02-24T00:16:00Z</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             CreateDate  nSamples\n",
       "0  2019-02-23T22:52:00Z        16\n",
       "1  2019-02-23T23:37:00Z        16\n",
       "2  2019-02-23T22:58:00Z        15\n",
       "3  2019-02-23T23:01:00Z        15\n",
       "4  2019-02-24T00:16:00Z        13"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# batch samples based on SRA create date field\n",
    "batches = meta[\"CreateDate\"].value_counts().to_frame(\"nSamples\")\n",
    "batches.reset_index(drop=False, inplace = True)\n",
    "batches.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CreateDateBatch\n",
       "2019-02-23    542\n",
       "2019-02-24    456\n",
       "2019-02-25     20\n",
       "                1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# clip off time stamp, so we can use day\n",
    "# remove last 10 chr ex, T23:37:00Z\n",
    "# there will be three batches of data based on date w/o time stamp\n",
    "meta.insert(36, \"CreateDateBatch\", meta['CreateDate'].map(lambda x: str(x)[:-10]))\n",
    "meta['CreateDateBatch'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "# one sample does not have a batch, drop this sample\n",
    "meta = meta.loc[meta[\"CreateDateBatch\"] != \"\", ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a pseudo flow cell ID tag for sample\n",
    "# to mirror real world processes & approaches to binning NGS samples\n",
    "meta.insert(1, \"FlowCellID\", meta[\"CreateDateBatch\"].str.replace(\"-\", \"\") + \"_\" + meta[\"BioProject\"] + \"_\" + meta[\"Run\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3.2 Subsample samples within each batch\n",
    "\n",
    "We will only use a subset of samples per batch, for a total of 33 radomly selected samples, for the demo.\n",
    "\n",
    "1. Batch 2019-02-23 (n=8 samples)\n",
    "2. Batch 2019-02-24 (n=6 samples)\n",
    "3. Batch 2019-02-25 (n=19 samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">>>> 2019-02-23\n",
      ">>>> 2019-02-24\n",
      ">>>> 2019-02-25\n"
     ]
    }
   ],
   "source": [
    "# randomly subsample to pre-specified # of samples per batch within each batch\n",
    "\n",
    "n_samples_per_batch = {\"2019-02-23\" : 8,\n",
    "                       \"2019-02-24\" : 6,\n",
    "                       \"2019-02-25\" : 19\n",
    "                       }\n",
    "\n",
    "sub_meta = dict()\n",
    "for batch in n_samples_per_batch:\n",
    "    print(\">>>> \" + str(batch))\n",
    "    df = meta.query(\"CreateDateBatch == @batch\").sample(n_samples_per_batch[batch])\n",
    "    sub_meta[batch] = df\n",
    "sub_meta = pd.concat(sub_meta, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save formatted metadata for demo samples to disk\n",
    "sub_meta.to_csv(\"./demo_data/sample_metadata/demo_ccle_rnaseq_metadata.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4. Download fastqs\n",
    "\n",
    "After defining our cohort of interest, we can now download raw RNA-sequencing data for the demo samples of interest. Fastqs are downloaded directly from the Sequence Read Archive (SRA) using the [SRA TookKit (v3.0.6)](https://github.com/ncbi/sra-tools/wiki) command `fastq-dump`.\n",
    "\n",
    "We will use a simple bash script to perform the fastq download: \n",
    "\n",
    "\n",
    "\n",
    "```bash\n",
    "\n",
    "#!/bin/bash\n",
    "\n",
    "# define location of fastqs to be downloaded to\n",
    "cd ~/ccle_demo_fastqs/\n",
    "\n",
    "# for each sample in demo metadata, download fastqs & gzip them\n",
    "SAMPLES=$(tail -n +2 ./demo_data/sample_metadata/demo_ccle_rnaseq_metadata.csv | awk -F\",\" '{print $2}')\n",
    "for SRA in $SAMPLES:\n",
    "do\n",
    "    echo $SRA\n",
    "    prefetch -v $SRA\n",
    "    fastq-dump --outdir ./fastq/ --split-3 ./$SRA\n",
    "done;\n",
    "\n",
    "# gzip fastqs to save space (this is standard practice)\n",
    "# gzip compression level ranges from 1 to 9; level 1 is fastest, level 6 is default\n",
    "gzip -3 ~/ccle_demo_fastqs/SRR*\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we have the raw RNA-sequencing data, we are ready to proceed to the next step - defining a metadata schema and Quilt workflow to ensure the integrity of the metadata when attached to a Quilt package!\n",
    "\n",
    "----"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
